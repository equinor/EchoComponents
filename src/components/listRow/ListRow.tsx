import { Typography } from '@equinor/eds-core-react';
import React, { CSSProperties } from 'react';
import { themeConst } from '../..';
import { Icon } from '../../elements/icon/Icon';
import { ExpandableRowProps } from '../../types/expandableRowProps';
import { IconListItem } from '../../types/iconListItem';
import styles from './listRow.module.css';

export interface ListRowProps {
    isDraggable: boolean;
    item: IconListItem;
    expandable?: ExpandableRowProps;
    rowIndex: number;
    style?: CSSProperties;
}

export const ListRow: React.FC<ListRowProps> = ({
    isDraggable,
    item,
    expandable,
    style
}: ListRowProps): JSX.Element => {
    return (
        <div className={styles.wrapper} style={style}>
            <div className={styles.row}>
                <div className={styles.infoCol}>
                    {isDraggable && (
                        <div className={styles.draggable}>
                            <Icon name="unfold_more" title="Move" color={themeConst.asBuilt}></Icon>
                        </div>
                    )}
                    <div className={styles.info}>
                        {item.title && (
                            <Typography variant="body_short" bold>
                                {item.title}
                            </Typography>
                        )}
                        {item.subTitle && <Typography variant="body_short">{item.subTitle}</Typography>}
                    </div>
                </div>
                <div className={styles.iconCol}>
                    {item.icons.map((icon, index2) => {
                        return (
                            <button
                                className={styles.iconButton}
                                key={index2}
                                onClick={(): void => {
                                    icon.onClick();
                                }}
                            >
                                <Icon name={icon.icon} title={icon.icon} color={themeConst.asBuilt}></Icon>
                            </button>
                        );
                    })}
                    {expandable && (
                        <button
                            className={expandable && expandable.expanded ? styles.activeIconButton : styles.iconButton}
                            onClick={(): void => {
                                if (expandable) {
                                    expandable.setExpanded(true);
                                }
                            }}
                        >
                            <Icon name="more_vertical" title="Expand" color={themeConst.asBuilt}></Icon>
                        </button>
                    )}
                </div>
            </div>
            {expandable && expandable.expanded && (
                <div className={styles.expandedRow}>
                    {expandable.iconItems.map((icon, index3) => {
                        return (
                            <button
                                className={styles.iconButton}
                                key={index3}
                                onClick={(): void => {
                                    icon.onClick();
                                }}
                            >
                                <Icon name={icon.icon} title={icon.icon} color={themeConst.asBuilt}></Icon>
                            </button>
                        );
                    })}
                    <button
                        className={styles.iconButton}
                        onClick={(): void => {
                            if (expandable) {
                                expandable.setExpanded(false);
                            }
                        }}
                    >
                        <Icon name={'close'} title={'Close'} color={themeConst.asBuilt}></Icon>
                    </button>
                </div>
            )}
        </div>
    );
};
