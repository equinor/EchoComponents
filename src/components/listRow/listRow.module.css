.wrapper {
    border-bottom: 1px solid #dcdcdc;
    display: flex;
    flex-direction: column;
}

.row {
    display: flex;
    justify-content: space-between;
    min-height: 48px;
    align-items: stretch;
}

.infoCol {
    display: flex;
    margin: auto 0;
}

.draggable {
    display: flex;
    align-items: center;
    cursor: pointer;
}

.info {
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-left: var(--medium);
}

.info p {
    font-size: 14px;
}

.iconCol {
    display: flex;
    align-items: stretch;
}

.activeIconButton {
    align-items: center;
    display: flex;
    padding: 0 12px;
    background-color: var(--equiGreen3);
}

.iconButton {
    align-items: center;
    display: flex;
    padding: 0 12px;
}

.iconButton:hover {
    background-color: #c3f3d2;
}

.iconButton:active,
.iconButton:focus {
    background-color: var(--equiGreen3);
}

.expandedRow {
    display: flex;
    justify-content: flex-end;
    min-height: 48px;
    align-items: stretch;
    background-color: var(--equiGreen3);
}

.close {
    width: 54px;
    justify-content: center;
    align-items: center;
    display: flex;
    cursor: pointer;
}
