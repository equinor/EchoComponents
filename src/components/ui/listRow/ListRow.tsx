import { themeConst } from '@equinor/echo-framework';
import { Typography } from '@equinor/eds-core-react';
import cx from 'classnames';
import React, { CSSProperties } from 'react';
import { DraggableHandleSelector } from '../../../elements/functional/draggableOrder/DraggableOrder';
import { Icon } from '../../../elements/icons/icon/Icon';
import styles from './listRow.module.css';

export interface ListRowProps {
    isMovable: boolean;
    item: IconListItem;
    expandedIcons: IconItem[];
    expanded?: boolean;
    rowIndex: number;
    onExpand?: (status: boolean) => void;
    style?: CSSProperties;
}

export interface IconListItem {
    title?: string;
    subTitle?: string;
    icons: IconItem[];
}

export interface IconItem {
    icon: string;
    onClick: () => void;
}

export const ListRow: React.FC<ListRowProps> = ({
    isMovable,
    item,
    expandedIcons,
    expanded,
    onExpand,
    style
}: ListRowProps): JSX.Element => {
    return (
        <div className={styles.listRow} style={style}>
            <div className={styles.row1}>
                <div className={styles.col1}>
                    {isMovable && (
                        <div className={cx(styles.moveable, DraggableHandleSelector)}>
                            <Icon name="unfold_more" title="Move" color={themeConst.asBuilt}></Icon>
                        </div>
                    )}
                    <div className={cx(styles.info, isMovable ? '' : styles.infoMargin)}>
                        {item.title && (
                            <Typography variant="body_short" bold>
                                {item.title}
                            </Typography>
                        )}
                        {item.subTitle && <Typography variant="body_short">{item.subTitle}</Typography>}
                    </div>
                </div>
                <div className={styles.col2}>
                    {item.icons.map((icon, index2) => {
                        return (
                            <div
                                className={styles.iconWrapper}
                                key={index2}
                                onClick={(): void => {
                                    icon.onClick();
                                }}
                                tabIndex={0}
                            >
                                <Icon name={icon.icon} title={icon.icon} color={themeConst.asBuilt}></Icon>
                            </div>
                        );
                    })}
                    {expandedIcons && (
                        <div
                            className={styles.expandable}
                            onClick={(): void => {
                                if (onExpand) {
                                    onExpand(true);
                                }
                            }}
                            tabIndex={0}
                        >
                            <Icon name="more_vertical" title="Expand" color={themeConst.asBuilt}></Icon>
                        </div>
                    )}
                </div>
            </div>
            {expanded && (
                <div className={styles.row2}>
                    {expandedIcons.map((icon, index3) => {
                        return (
                            <div
                                className={styles.iconWrapper}
                                key={index3}
                                onClick={(): void => {
                                    icon.onClick();
                                }}
                                tabIndex={0}
                            >
                                <Icon name={icon.icon} title={icon.icon} color={themeConst.asBuilt}></Icon>
                            </div>
                        );
                    })}
                    <div
                        className={styles.close}
                        onClick={(): void => {
                            if (onExpand) {
                                onExpand(false);
                            }
                        }}
                        tabIndex={0}
                    >
                        <Icon name={'close'} title={'Close'} color={themeConst.asBuilt}></Icon>
                    </div>
                </div>
            )}
        </div>
    );
};
