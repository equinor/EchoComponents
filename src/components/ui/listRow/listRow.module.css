.listRow {
    border-bottom: 1px solid #dcdcdc;
    display: flex;
    flex-direction: column;
}

.row1 {
    display: flex;
    justify-content: space-between;
    height: 100%;
    align-items: stretch;
}

.col1 {
    display: flex;
    padding: 10px 0;
}

.moveable {
    margin: 0 10px 0 10px;
    display: flex;
    align-items: center;
    cursor: pointer;
}

.info {
    display: flex;
    flex-direction: column;
    justify-content: center;
}

.infoMargin {
    margin: 0 0 0 20px;
}

.info p {
    font-size: 14px;
}
.col2 {
    display: flex;
    align-items: stretch;
}

.iconWrapper {
    cursor: pointer;
    align-items: center;
    display: flex;
    width: 54px;
    justify-content: center;
}

.expandable {
    width: 54px;
    align-items: center;
    display: flex;
    cursor: pointer;
    justify-content: center;
}

.expandable:hover {
    background-color: #e6faec;
}

.row2 {
    display: flex;
    justify-content: flex-end;
    height: 54px;
    align-items: stretch;
    background-color: #e6faec;
}

.close {
    width: 54px;
    justify-content: center;
    align-items: center;
    display: flex;
    cursor: pointer;
}
