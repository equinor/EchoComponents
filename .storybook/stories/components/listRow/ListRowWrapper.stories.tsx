import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import ListRowWrapper, { ListRowWrapperProps } from './ListRowWrapper';

export default {
    title: 'Components/ListRow',
    component: ListRowWrapper,
    argTypes: {
        expandedIcons: {
            control: {
                type: 'inline-check',
                options: ['cloud_upload', 'file_copy', 'edit', 'link', 'delete_to_trash', 'sync']
            }
        },
        icons: {
            control: {
                type: 'inline-check',
                options: ['cloud_upload', 'file_copy', 'edit', 'link', 'delete_to_trash', 'sync']
            }
        }
    }
} as Meta;

const Template: Story<ListRowWrapperProps> = (args) => <ListRowWrapper {...args} />;

export const Default = Template.bind({});

Default.args = {
    style: { width: '350px' },
    isDraggable: true,
    title: 'Item 0',
    subTitle: 'SubItem 0',
    icons: ['link', 'delete_to_trash'],
    expandedIcons: ['cloud_upload'],
    expandable: true
};
