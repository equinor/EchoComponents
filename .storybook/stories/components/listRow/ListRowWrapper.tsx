import React, { CSSProperties, useState } from 'react';
import { ListRow } from '../../../../src/components/listRow/ListRow';

export interface FormGeneratorWrapperProps {
    style: CSSProperties;
    isDraggable: boolean;
    title: string;
    subTitle: string;
    icons: string[];
    expandedIcons: string[];
    expandable: boolean;
}

const FormGeneratorWrapper: React.FC<FormGeneratorWrapperProps> = ({
    isDraggable,
    title,
    subTitle,
    icons,
    expandedIcons,
    expandable,
    style
}: FormGeneratorWrapperProps) => {
    const [open, setOpen] = useState<boolean>(false);

    return (
        <div className={'ListRow'}>
            <ListRow
                isDraggable={isDraggable}
                item={{
                    title,
                    subTitle,
                    icons: icons.map((i) => {
                        return {
                            icon: i,
                            onClick: () => {
                                console.log('clicked');
                            }
                        };
                    })
                }}
                expandable={
                    expandable
                        ? {
                              iconItems: expandedIcons.map((i) => {
                                  return {
                                      icon: i,
                                      onClick: () => {
                                          console.log('clicked');
                                      }
                                  };
                              }),
                              expanded: open,
                              setExpanded: (status: boolean) => {
                                  setOpen(status);
                              }
                          }
                        : undefined
                }
                rowIndex={0}
                style={style}
            ></ListRow>
        </div>
    );
};

export default FormGeneratorWrapper;
