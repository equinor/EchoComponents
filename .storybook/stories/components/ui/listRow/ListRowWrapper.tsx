import React, { useState } from 'react';
import { ListRow, ListRowProps } from '../../../../../src/components/ui/listRow/ListRow';

const FormGeneratorWrapper: React.FC<ListRowProps> = ({
    isMovable,
    item,
    expandedIcons,
    rowIndex,
    expanded,
    onExpand,
    style
}: ListRowProps) => {
    const [open, setOpen] = useState<boolean>(expanded);

    onExpand = (status: boolean) => {
        setOpen(status);
    };

    return (
        <div className={'ListRow'}>
            <ListRow
                isMovable={isMovable}
                item={item}
                expandedIcons={expandedIcons}
                rowIndex={rowIndex}
                expanded={open}
                onExpand={onExpand}
                style={style}
            ></ListRow>
        </div>
    );
};

export default FormGeneratorWrapper;
