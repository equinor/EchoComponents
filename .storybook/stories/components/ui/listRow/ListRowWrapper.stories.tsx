import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import { ListRowProps } from '../../../../../src/components/ui/listRow/ListRow';
import ListRowWrapper from './ListRowWrapper';

export default {
    title: 'Components/UI/ListRow',
    component: ListRowWrapper,
    argTypes: {}
} as Meta;

const Template: Story<ListRowProps> = (args) => <ListRowWrapper {...args} />;

export const Default = Template.bind({});

Default.args = {
    style: { width: '350px' },
    isMovable: true,
    item: {
        title: 'Item 0',
        subTitle: 'SubItem 0',
        icons: [
            {
                icon: 'link',
                onClick: () => {}
            },
            {
                icon: 'delete_to_trash',
                onClick: () => {}
            }
        ]
    },
    expandedIcons: [
        {
            icon: 'cloud_upload',
            onClick: () => {}
        },
        {
            icon: 'file_copy',
            onClick: () => {}
        },
        {
            icon: 'edit',
            onClick: () => {}
        }
    ],
    expanded: false,
    rowIndex: 0,
    onExpand: () => {}
};
