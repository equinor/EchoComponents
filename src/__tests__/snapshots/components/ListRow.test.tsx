import React from 'react';
import renderer from 'react-test-renderer';
import { ListRow } from '../../../components/listRow/ListRow';

it('renders ListRow correctly without expandable row', () => {
    const tree = renderer
        .create(
            <ListRow
                isDraggable={true}
                item={{
                    title: 'First item',
                    subTitle: 'Sub title',
                    icons: [{ icon: 'edit', onClick: jest.fn() }]
                }}
                rowIndex={0}
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders ListRow correctly with expandable row', () => {
    const tree = renderer
        .create(
            <ListRow
                isDraggable={true}
                item={{
                    title: 'First item',
                    subTitle: 'Sub title',
                    icons: [{ icon: 'edit', onClick: jest.fn() }]
                }}
                rowIndex={0}
                expandable={{
                    expanded: true,
                    setExpanded: jest.fn(),
                    iconItems: [{ icon: 'edit', onClick: jest.fn() }]
                }}
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
