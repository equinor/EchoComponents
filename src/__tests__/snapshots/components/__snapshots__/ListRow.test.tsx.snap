// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ListRow correctly with expandable row 1`] = `
<div
  className="wrapper"
>
  <div
    className="row"
  >
    <div
      className="infoCol"
    >
      <div
        className="draggable"
      >
        <svg
          aria-labelledby="eds-unfold_more-3"
          className="sc-iwyYcG dqcEJR"
          fill="#007079"
          height="24px"
          name={null}
          role="img"
          title="Move"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="eds-unfold_more-3"
          >
            Move
          </title>
          <path
            className="sc-cxFLnm"
            clipRule="evenodd"
            d="M15.17 9L12 5.83 8.83 9 7.41 7.59 12 3l4.58 4.59L15.17 9zm-6.34 6L12 18.17 15.17 15l1.42 1.41L12 21l-4.58-4.59L8.83 15z"
            data-testid="eds-icon-path"
            fillRule="evenodd"
            height="24"
            size={null}
            transform={null}
          />
        </svg>
      </div>
      <div
        className="info"
      >
        <p
          className="sc-eCssSg fvZGxZ"
        >
          First item
        </p>
        <p
          className="sc-eCssSg bHjGEK"
        >
          Sub title
        </p>
      </div>
    </div>
    <div
      className="iconCol"
    >
      <button
        className="iconButton"
        onClick={[MockFunction]}
      >
        <svg
          aria-labelledby="eds-edit-4"
          className="sc-iwyYcG dqcEJR"
          fill="#007079"
          height="24px"
          name={null}
          role="img"
          title="edit"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="eds-edit-4"
          >
            edit
          </title>
          <path
            className="sc-cxFLnm"
            clipRule="evenodd"
            d="M18.368 3.29l2.34 2.34c.39.39.39 1.02 0 1.41l-1.83 1.83-3.75-3.75 1.83-1.83c.19-.19.44-.29.7-.29.26 0 .51.09.71.29zM2.998 17.25V21h3.75l11.06-11.06-3.75-3.75-11.06 11.06zM5.918 19h-.92v-.92l9.06-9.06.92.92L5.918 19z"
            data-testid="eds-icon-path"
            fillRule="evenodd"
            height="24"
            size={null}
            transform={null}
          />
        </svg>
      </button>
      <button
        className="activeIconButton"
        onClick={[Function]}
      >
        <svg
          aria-labelledby="eds-more_vertical-5"
          className="sc-iwyYcG dqcEJR"
          fill="#007079"
          height="24px"
          name={null}
          role="img"
          title="Expand"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="eds-more_vertical-5"
          >
            Expand
          </title>
          <path
            className="sc-cxFLnm"
            clipRule="evenodd"
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-2 8c0-1.1.9-2 2-2s2 .9 2 2-.9 2-2 2-2-.9-2-2z"
            data-testid="eds-icon-path"
            fillRule="evenodd"
            height="24"
            size={null}
            transform={null}
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="expandedRow"
  >
    <button
      className="iconButton"
      onClick={[MockFunction]}
    >
      <svg
        aria-labelledby="eds-edit-6"
        className="sc-iwyYcG dqcEJR"
        fill="#007079"
        height="24px"
        name={null}
        role="img"
        title="edit"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="eds-edit-6"
        >
          edit
        </title>
        <path
          className="sc-cxFLnm"
          clipRule="evenodd"
          d="M18.368 3.29l2.34 2.34c.39.39.39 1.02 0 1.41l-1.83 1.83-3.75-3.75 1.83-1.83c.19-.19.44-.29.7-.29.26 0 .51.09.71.29zM2.998 17.25V21h3.75l11.06-11.06-3.75-3.75-11.06 11.06zM5.918 19h-.92v-.92l9.06-9.06.92.92L5.918 19z"
          data-testid="eds-icon-path"
          fillRule="evenodd"
          height="24"
          size={null}
          transform={null}
        />
      </svg>
    </button>
    <button
      className="iconButton"
      onClick={[Function]}
    >
      <svg
        aria-labelledby="eds-close-7"
        className="sc-iwyYcG dqcEJR"
        fill="#007079"
        height="24px"
        name={null}
        role="img"
        title="Close"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="eds-close-7"
        >
          Close
        </title>
        <path
          className="sc-cxFLnm"
          clipRule="evenodd"
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          data-testid="eds-icon-path"
          fillRule="evenodd"
          height="24"
          size={null}
          transform={null}
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`renders ListRow correctly without expandable row 1`] = `
<div
  className="wrapper"
>
  <div
    className="row"
  >
    <div
      className="infoCol"
    >
      <div
        className="draggable"
      >
        <svg
          aria-labelledby="eds-unfold_more-1"
          className="sc-iwyYcG dqcEJR"
          fill="#007079"
          height="24px"
          name={null}
          role="img"
          title="Move"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="eds-unfold_more-1"
          >
            Move
          </title>
          <path
            className="sc-cxFLnm"
            clipRule="evenodd"
            d="M15.17 9L12 5.83 8.83 9 7.41 7.59 12 3l4.58 4.59L15.17 9zm-6.34 6L12 18.17 15.17 15l1.42 1.41L12 21l-4.58-4.59L8.83 15z"
            data-testid="eds-icon-path"
            fillRule="evenodd"
            height="24"
            size={null}
            transform={null}
          />
        </svg>
      </div>
      <div
        className="info"
      >
        <p
          className="sc-eCssSg fvZGxZ"
        >
          First item
        </p>
        <p
          className="sc-eCssSg bHjGEK"
        >
          Sub title
        </p>
      </div>
    </div>
    <div
      className="iconCol"
    >
      <button
        className="iconButton"
        onClick={[MockFunction]}
      >
        <svg
          aria-labelledby="eds-edit-2"
          className="sc-iwyYcG dqcEJR"
          fill="#007079"
          height="24px"
          name={null}
          role="img"
          title="edit"
          viewBox="0 0 24 24"
          width="24px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="eds-edit-2"
          >
            edit
          </title>
          <path
            className="sc-cxFLnm"
            clipRule="evenodd"
            d="M18.368 3.29l2.34 2.34c.39.39.39 1.02 0 1.41l-1.83 1.83-3.75-3.75 1.83-1.83c.19-.19.44-.29.7-.29.26 0 .51.09.71.29zM2.998 17.25V21h3.75l11.06-11.06-3.75-3.75-11.06 11.06zM5.918 19h-.92v-.92l9.06-9.06.92.92L5.918 19z"
            data-testid="eds-icon-path"
            fillRule="evenodd"
            height="24"
            size={null}
            transform={null}
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
